### Q1. What is the difference between Ordinal Encoding and Label Encoding? Provide an example of when you might choose one over the other.

**Ordinal Encoding** is a technique where the categories of a categorical variable are assigned an ordered numerical value based on their rank or position. 

> For example, if we have a variable "education level" with categories "High School", "College", and "Graduate School", we can assign them values 1, 2, and 3, respectively. This technique assumes an inherent order or hierarchy in the categories and is useful when the categories have a clear ordering, such as in the case of education level, where "Graduate School" is higher than "College", which in turn is higher than "High School".

**Label Encoding**, on the other hand, is a technique where each category of a categorical variable is assigned a unique numerical value. 

> For example, if we have a variable "color" with categories "Red", "Green", and "Blue", we can assign them values 1, 2, and 3, respectively. This technique does not assume any order or hierarchy in the categories and is useful when the categories do not have an inherent order.

**An example where we might choose one over the other** is when encoding the categorical variable "size" with categories "Small", "Medium", and "Large". If we believe that the size categories have a clear ordering, such as "Small" being smaller than "Medium" and "Medium" being smaller than "Large", we might choose to use Ordinal Encoding. On the other hand, if we believe that the size categories do not have an inherent order, such as "Red", "Green", and "Blue" in the "color" example, we might choose to use Label Encoding.

### Q2. Explain how Target Guided Ordinal Encoding works and provide an example of when you might use it in a machine learning project.


The steps for Target Guided Ordinal Encoding are as follows:
1. Calculate the **mean of the target variable for each category** of the categorical variable.
2. **Order the categories based on the mean of the target variable**, with the category having the highest mean assigned the highest ordinal value.
3. **Assign an ordinal value to each category** based on its rank in the ordered list.

**For example**, let's say we have a categorical variable `"city"` with categories `"New York"`, `"San Francisco"`, and `"Seattle"`, and we want to predict a target variable `"income"`. We can perform Target Guided Ordinal Encoding as follows:

1. Calculate the mean income for each city: "New York" has a mean income of 75,000 dollars "San Francisco" has a mean income of 100,000 dollars and "Seattle" has a mean income of 85,000 dollars.
2. Order the cities based on the mean income: "San Francisco" has the highest mean income, followed by "Seattle", and then "New York".
3. Assign an ordinal value to each city based on its rank in the ordered list: "San Francisco" is assigned the ordinal value of 3, "Seattle" is assigned the ordinal value of 2, and "New York" is assigned the ordinal value of 1.

### Q3. Define covariance and explain why it is important in statistical analysis. How is covariance calculated?


**Covariance** is a measure of the extent to which two random variables are linearly related. Specifically, it measures the degree to which the values of one variable change in relation to the values of another variable.

**If two variables have a positive covariance,** it means that when one variable increases, the other variable tends to increase as well.  
**If they have a negative covariance,** it means that when one variable increases, the other variable tends to decrease.  
**If the covariance is zero,** it means that the variables are not linearly related.

**Covariance is important in statistical analysis** because it is a measure of the strength and direction of the relationship between two variables. This relationship can provide important insights into the underlying nature of the data and can inform decisions about modeling and prediction. For example, in finance, covariance is used to measure the degree to which the returns on two different stocks are related, which can help investors diversify their portfolios.

**Covariance is calculated** as the sum of the product of the deviations of each variable from its mean, divided by the sample size minus one. The formula for the covariance between two variables X and Y with sample size n is:

> $cov(X,Y) = 1/(n-1) * ∑(X_i - \bar{X}) * (Y_i - \bar{Y})$

where $X_i$ and $Y_i$ are the values of the variables at the $i^{th}$ observation, $\bar{X}$ and $\bar{Y}$ are the means of the variables, and the summation is taken over all $n$ observations.

### Q4. For a dataset with the following categorical variables: Color (red, green, blue), Size (small, medium, large), and Material (wood, metal, plastic), perform label encoding using Python's scikit-learn library. Show your code and explain the output.


In [ ]:
import pandas as pd

df = pd.DataFrame({
    'Color': ['Red', 'Green', 'Blue'],
    'Size': ['Small', 'Medium', 'Large'],
    'Material': ['Wood', 'Metal', 'Plastic']
})

df.head()

Color    Size Material
0    Red   Small     Wood
1  Green  Medium    Metal
2   Blue   Large  Plastic

In [ ]:
from sklearn.preprocessing import LabelEncoder
encoder = LabelEncoder()

color_encoded = encoder.fit_transform(df['Color'])
size_encoded = encoder.fit_transform(df['Size'])
material_encoded = encoder.fit_transform(df['Material'])

df_encoded = pd.DataFrame({
    'Color_Enc': color_encoded.tolist(),
    'Size_Enc': size_encoded.tolist(),
    'Material_Enc': material_encoded.tolist(),
})

df_encoded

Color_Enc  Size_Enc  Material_Enc
0          2         2             2
1          1         1             0
2          0         0             1

The resulting encoded values are simply numerical representations of the corresponding categorical values. In other words, they are just arbitrary integers assigned to each category. These values do not have any inherent meaning or order, and they do not reflect any underlying relationship between the categories.

### Q5. Calculate the covariance matrix for the following variables in a dataset: Age, Income, and Education level. Interpret the results.


In [ ]:
import numpy as np

age_array = np.array([32, 24, 52, 43])
income_array = np.array([60000, 45000, 75000, 70000])
education_array = np.array([5, 4, 6, 8])

data_matrix = np.column_stack((age_array, income_array, education_array))

covariance_matrix = np.cov(data_matrix.T)

covariance_matrix

array([[1.50916667e+02, 1.57500000e+05, 1.45833333e+01],
       [1.57500000e+05, 1.75000000e+08, 1.75000000e+04],
       [1.45833333e+01, 1.75000000e+04, 2.91666667e+00]])

**Interpretation of the results:**

The covariance between age and income is 28,416.67, indicating that there is a positive relationship between these two variables. This means that as age increases, income tends to increase as well.  
The covariance between age and education level is 166.67, indicating a weak positive relationship. This means that as age increases, education level tends to increase slightly.  
The covariance between income and education level is 4,666.67, indicating a weak positive relationship. This means that as income increases, education level tends to increase slightly.


### Q6. You are working on a machine learning project with a dataset containing several categorical variables, including "Gender" (Male/Female), "Education Level" (High School/Bachelor's/Master's/PhD), and "Employment Status" (Unemployed/Part-Time/Full-Time). Which encoding method would you use for each variable, and why?


For the "Gender" variable, I would use Label Encoding, as there are only two categories, Male and Female, and there is no inherent order or hierarchy between the categories.

For the "Education Level" variable, I would use Ordinal Encoding, as there is an inherent order or hierarchy between the categories, with a higher education level being "better" than a lower education level. I would assign an ordinal value to each category based on its level of education, with High School being the lowest and PhD being the highest.

For the "Employment Status" variable, I would use One-Hot Encoding, as there are three categories, and there is no inherent order or hierarchy between the categories. One-Hot Encoding creates a binary variable for each category, with a value of 1 if the observation belongs to that category, and a value of 0 otherwise. This approach ensures that there is no implied ranking or order between the categories.

### Q7. You are analyzing a dataset with two continuous variables, "Temperature" and "Humidity", and two categorical variables, "Weather Condition" (Sunny/Cloudy/Rainy) and "Wind Direction" (North/South/East/West). Calculate the covariance between each pair of variables and interpret the results.

To calculate the covariance between each pair of variables, we need to compute the deviations of each variable from their respective means and multiply them together, then take the average of the resulting products over all observations. The formula for the sample covariance between two variables X and Y is:

> $cov(X,Y) = (1/n-1) * ∑(X_i - \bar{X}) * (Y_i - \bar{Y})$

where $X_i$ and $Y_i$ are the values of the variables at the ith observation, $\bar{X}$ and $\bar{Y}$ are the means of the variables, and the summation is taken over all $n$ observations.  
We can also use the pandas cov function to find the covariance.

Assuming we have the data for all variables, we can calculate the covariance between each pair of the variables as follows:

In [ ]:
import pandas as pd

df = pd.DataFrame({
    'Temperature': [40,22,25,45,36],
    'Humidity': [60,85,40,56,22],
    'Weather Condition': ['Sunny', 'Cloudy', 'Rainy','Sunny', 'Cloudy'],
    'Wind Direction': ['North', 'South', 'East','South', 'West']
})

df.head()

Temperature  Humidity Weather Condition Wind Direction
0           40        60             Sunny          North
1           22        85            Cloudy          South
2           25        40             Rainy           East
3           45        56             Sunny          South
4           36        22            Cloudy           West

Since Weather Condition and Wind Direction are categorical variables, we cannot directly compute the covariance between them. Instead, we would need to convert them into numeric variables using an encoding method such as one-hot encoding or ordinal encoding. 

In [ ]:
from sklearn.preprocessing import LabelEncoder
encoder = LabelEncoder()

weather_encoded = encoder.fit_transform(df['Weather Condition'])
wind_encoded = encoder.fit_transform(df['Wind Direction'])

df_encoded = pd.DataFrame({
    'Weather Cond Encoded': weather_encoded.tolist(),
    'Wind Dir Encoded': wind_encoded.tolist()
})

df_new = pd.concat([df, df_encoded], axis=1)
df_new = df_new.drop('Weather Condition', axis=1)
df_new = df_new.drop('Wind Direction', axis=1)

df_new

Temperature  Humidity  Weather Cond Encoded  Wind Dir Encoded
0           40        60                     2                 1
1           22        85                     0                 2
2           25        40                     1                 0
3           45        56                     2                 2
4           36        22                     0                 3

In [ ]:
# covariance between Temperature and Humidity
cov_1 = df_new['Temperature'].cov(df_new['Humidity'])

# covariance between Weather Condition and Wind Direction
cov_2 = df_new['Weather Cond Encoded'].cov(df_new['Wind Dir Encoded'])

print(f"Covariance between Temperature and Humidity: {cov_1}")
print(f"Covariance between Weather Condition and Wind Direction: {cov_2}")

Covariance between Temperature and Humidity: -63.7
Covariance between Weather Condition and Wind Direction: -0.5


In the given results, we have:

**Covariance between Temperature and Humidity: -63.7**
The negative value indicates that as temperature increases, humidity tends to decrease and vice versa. This may suggest an inverse relationship between temperature and humidity. For example, on a hot day, humidity is likely to be low.

**Covariance between Weather Condition and Wind Direction: -0.5**
Again, the negative value indicates an inverse relationship, but the small magnitude suggests that the two variables are not strongly related. This covariance result may indicate that changes in wind direction do not have a significant impact on weather conditions or vice versa.

[Source Notebook](https://github.com/liskibruh/data-sciences-masters-course-notebooks/blob/main/Day048_Assignment.ipynb)